go/libraries/doltcore/sqle/auto_gc.go

```diff
@@ -0,0 +1,348 @@
+	// we can try to send when we request a GC. If will become our
+	// new |done| channel once we send it successfully.
+	next chan struct{}
+	// lastSz is set the first time we observe StoreSizes after a
+	// GC or after the server comes up. It is used in some simple
+	// growth heuristics to figure out if we want to run a GC. We
+	// set it back to |nil| when we successfully submit a request
+	// to GC, so that we observe and store the new size after the
+	// GC is finished.
+	lastSz *doltdb.StoreSizes
+
+	db *doltdb.DoltDB
+
+	// Closed when the thread should shutdown because the database
+	// is being removed.
+	stopCh chan struct{}
+	// An optimistic send on this channel notifies the background
+	// thread that the sizes may have changed and it can check for
+	// the GC condition.
+	tickCh chan struct{}
+	wg     sync.WaitGroup
+}
+
+// During engine initialization, called on the original set of
+// databases to configure them for auto-GC.
+func (c *AutoGCController) ApplyCommitHooks(ctx context.Context, mrEnv *env.MultiRepoEnv, dbs ...dsess.SqlDatabase) error {
+	for _, db := range dbs {
+		denv := mrEnv.GetEnv(db.Name())
+		if denv == nil {
+			continue
+		}
+		ddb := denv.DoltDB(ctx)
+		ddb.PrependCommitHooks(ctx, c.newCommitHook(db.Name(), ddb))
+	}
+	return nil
+}
+
+func (c *AutoGCController) DropDatabaseHook() DropDatabaseHook {
+	return func(_ *sql.Context, name string) {
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		hook := c.hooks[name]
+		if hook != nil {
+			hook.stop()
+			delete(c.hooks, name)
+		}
+	}
+}
+
+func (c *AutoGCController) InitDatabaseHook() InitDatabaseHook {
+	return func(ctx *sql.Context, _ *DoltDatabaseProvider, name string, env *env.DoltEnv, _ dsess.SqlDatabase) error {
+		ddb := env.DoltDB(ctx)
+		ddb.PrependCommitHooks(ctx, c.newCommitHook(name, ddb))
+		return nil
+	}
+}
+
+func (h *autoGCCommitHook) Execute(ctx context.Context, _ datas.Dataset, _ *doltdb.DoltDB) (func(context.Context) error, error) {
+	select {
+	case h.tickCh <- struct{}{}:
+		return nil, nil
+	case <-ctx.Done():
+		return nil, context.Cause(ctx)
+	}
+}
+
+func (h *autoGCCommitHook) requestGC(ctx context.Context) error {
+	select {
+	case h.c.workCh <- autoGCWork{h.db, h.next, h.name}:
+		h.done = h.next
+		h.next = make(chan struct{})
+		h.lastSz = nil
+		return nil
+	case <-ctx.Done():
+		return context.Cause(ctx)
+	}
+}
+
+func (h *autoGCCommitHook) HandleError(ctx context.Context, err error) error {
+	return nil
+}
+
+func (h *autoGCCommitHook) SetLogger(ctx context.Context, wr io.Writer) error {
+	return nil
+}
+
+func (h *autoGCCommitHook) ExecuteForWorkingSets() bool {
+	return true
+}
+
+func (h *autoGCCommitHook) checkForGC(ctx context.Context) error {
+	select {
+	case <-h.done:
+		sz, err := h.db.StoreSizes(ctx)
+		if err != nil {
+			// Something is probably quite wrong. Regardless, can't determine if we should GC.
+			return err
+		}
+		if h.lastSz == nil {
+			h.lastSz = &sz
+		}
+		const size_128mb = (1 << 27)
```

Comment:
Threshold values probably should be defined as constants.

Might also consider making these settable via environment vars. A reasonable testing strategy might be running benchmarks / other heavy IO processes multithreaded with a quite low threshold and making sure perf is reasonable / no deadlock happen.

Reply:
Anything that needs to be tuneable is gonna be tuneable via config.yaml eventually, but this first pass just has them hard coded...

Same for the timer on the periodic check.